subsys: fs/nvs: Improvements

Laczen opened the pull request:
NVS needed improvements to provide better support for systems with write block size > 1 byte. Also the mechanism to protect NVS against incomplete writes was not good enough. This PR provides a rewrite of NVS to handle both problems.

The way data is stored on flash, has been modified: data is stored in length-id-data-length and this complete set is aligned. The length at the end is used as a verification at start (nvs_init). An advantage of this change is also that the metadata (length and id) takes up less space so more data can be stored.

A CRC is not added to the data, but (if required) this can be done by the user application easily.

After each write a read is performed to check if the data was written OK.

When an error is detected (either during write/verification/initialization) NVS goes into a locked state. This allows the recovery of data before the error. A call to reinit() when in locked state erases flash storage and enables write acces.

Laczen committed:
subsys: fs/nvs: Improved nvs for larger blocksizes

The nvs module has some disadvantages for larger block size. The data
header and slot are taking up to much space. A rewrite is proposed that
reduces the used storage space for systems with write block size > 4.

The data storage in flash is now one unit consisting of: data_length,
data_id, data and data_length again in a multiple of the write block
size. The data_length at the end is used to validate the correctness of
the flash write and also allows to travel backwards in the filesystem.

As a comparison, on a system with block size 8 byte, a 32 bit values
now fits 1 block including the metadata (length and id). This used to
be 3 blocks.

The data_length will occupy 1 byte if the data length is less than 128
byte, it will occupy 2 byte if the data length is 128 byte or more. The
data length is limited to 16383 byte.

Each write to flash is verified by a read back of the data.

The read performance is improved because reading is done backwards so
the latest items are found first.

When the filesystem is locked it can be unlocked by calling
reinit(), this will clear flash and setup everything for storage.

add sample documentation - README.rst

Update dtsi to include erase_block_size, use erase_block_size in sample

Update prj.conf to include CONFIG_MPU_ALLOW_FLASH_WRITE

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

Laczen commented:
Hi @aurel32, I have updated NVS to provide better support on systems with write block size > 1. If you have some time could you please test/review.

dbkinder commented on doc/subsystems/nvs/nvs.rst:
```diff
@@ -90,9 +97,14 @@ More generally then, with
+as::
 
-   SECTOR_COUNT * (SECTOR_SIZE-8) * PAGE_ERASES / (NS * (DS+8)) minutes
+   ``SECTOR_COUNT * (SECTOR_SIZE-8) * PAGE_ERASES / (NS * (DS+8)) minutes``
```
remove the double back ticks here.  The double colon at the end of the previous line says the following indented text is a code-block, so as written the back ticks will show up in the generated doc output.

dbkinder commented on samples/subsys/nvs/README.rst:
```diff
@@ -0,0 +1,23 @@
+Overview
+********
+
+A simple application demonstrating the use of the nvs module for Non-Volatile
```
Clean up some capitalization issues, and combine the sentences a bit for clarity:

    This is a simple application demonstrating use of the NVS
    module for non-volatile (flash) storage.  In this application,
    a counter is incremented on every reboot and stored in flash,
    the application reboots, and the reboot counter data is retrieved.

dbkinder commented on samples/subsys/nvs/README.rst:
```diff
@@ -0,0 +1,23 @@
+********************
+
+This sample can be found under :file:`samples/subsys/nvs/` in the
+Zephyr tree.
```
Missing the build and run commands (and sample output).

dbkinder commented on doc/subsystems/nvs/nvs.rst:
```diff
@@ -90,9 +97,14 @@ More generally then, with
-the expected device life (in minutes) can be calculated as::
+for DS < 128 byte, the expected device life (in minutes) can be calculated as::
+
+   ``SECTOR_COUNT * (SECTOR_SIZE-8) * PAGE_ERASES / (NS * (DS+4)) minutes``
```
remove the double back ticks here. The double colon at the end of the previous line says the following indented text is a code-block, so as written the back ticks will show up in the generated doc output.

dbkinder commented on doc/subsystems/nvs/nvs.rst:
```diff
@@ -75,13 +81,14 @@ equal to the pagesize (1024 bytes) and 2 sectors have been defined. Each
-for storage and will be erased after 168 minutes of system time.  With the
-expected device life of 20,000 writes, with two sectors writing every 168
-minutes, the device should last about 168*20,000 minutes, or about 6.5 years.
+bytes for the metadata (length: 1 byte, id: 2 bytes, length: 1 byte ), so in
```
Again, the two "length" metadata fields are confusing?

dbkinder commented on doc/subsystems/nvs/nvs.rst:
```diff
@@ -75,13 +81,14 @@ equal to the pagesize (1024 bytes) and 2 sectors have been defined. Each
+full.  When this happens, because we're using only two sectors, the first
+sector will be used for storage and will be erased after 254 minutes of system
+time.  With the expected device life of 20,000 writes, with two sectors writing
+every 254 minutes, the device should last about 254*20,000 minutes, or about
```
best to add a space around arithmetic operators:  ``254 * 20,000` minutes``

aurel32 commented:
Hi @Laczen, thanks for this new version. I'll try to have a look at it and try it tonight, if not tomorrow night. So far I have only seen it is based on an old version of the master branch which is in conflict with the current one. Any chance you can rebase it? Anyway that should not prevent me to test it.

nvlsianpu commented:
thanks for the patch - will review it soon.

Laczen commented:
Hi @dbkinder, I have updated the documentation as requested.

nashif commented on doc/subsystems/nvs/nvs.rst:
```diff
@@ -74,6 +96,7 @@ needs to be restored after reboot. NVS has been defined with a sector_size
 equal to the pagesize (1024 bytes) and 2 sectors have been defined. Each
 sector in NVS has a sector header of 8 bytes.
 
+<<<<<<< HEAD
```
please fix those, looks like a conflict that was not resolved.

Laczen commented:
@nashif, corrected

aurel32 commented:
Hi @Laczen. Thanks for the rebase, it makes the testing a lot more easy, as my application depends on changes that have been done recently.

I have tested your patch on my board, and it works as expected. I haven't been able to reproduce the issue with the garbage collector on power failure. I'll review the code itself tomorrow.

As said in our previous discussion, mixing the metadata and the data doesn't allow to implement a secure delete function (at least when the write block size is > 1), that said i believe we can do the same by triggering the garbage collector on the sector after writing the new/empty value.

nvlsianpu reviewed (changes requested):
will put much more commend on Monday

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -13,32 +13,41 @@
+/* basic calculation routines */
 
-static inline u16_t _nvs_len_in_flash(struct nvs_fs *fs, u16_t len)
+static size_t _nvs_flash_align(struct nvs_fs *fs, size_t len)
```
I will change this to: _get_wbs_aligned_len()

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -13,32 +13,41 @@
 }
 
-u16_t _nvs_entry_len_in_flash(struct nvs_fs *fs, u16_t len)
+size_t _nvs_flash_size(struct nvs_fs *fs, size_t len)
```
I will change this to get_wbs_aligned_entry_len()

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,104 +6,492 @@
+ */
+size_t _nvs_get_length(u8_t *buf, size_t *len)
+{
+	if (buf[0] & 0x80) {
```
There should still be a possibility to detect  0xff and 0xff. The range of 16383 bytes should be more then sufficient.

Laczen replied:
After rethinking, it is easily changed and easier to understand than the present. I will change it to the proposal, except length is max 0x7ffe.

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,104 +6,492 @@
+ * _nvs_get_length is used to determine if the flash is unwritten, therefor
+ * _nvs_get_length returns 0xffff if the data in buf[0] and buf[1] is 0xff.
+ */
+size_t _nvs_get_length(u8_t *buf, size_t *len)
```
Yes.

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,104 +6,492 @@
+ * _nvs_put_length returns the number of byte used and fills the buffer, the
+ * maximum allowed len is 16383 byte.
+ */
+size_t _nvs_put_length(u8_t *buf, size_t len)
```
Yes.

aurel32 commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,104 +6,492 @@
+int _nvs_flash_block_cmp(struct nvs_fs *fs, off_t offset, const void *data,
+		   size_t len)
+{
+	int rc, block_size;
```
You should define block_size as size_t, as it's the type returned by sizeof and also the type of len, so that the comparison is done on the same type.

Laczen replied:
Correct block_size should be size_t, will change

aurel32 commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,104 +6,492 @@
+	block_size = sizeof(buf);
+
+	while (len >= block_size) {
+		rc = _nvs_flash_read(fs, offset, buf, block_size);
```
I think the code can be simplified by using a `while (len > 0)`, and then computing bytes to compare as `min(block_size, len)`. Actually it's the strategy chosen a few lines before in` _nvs_flash_block_move`.

Laczen replied:
Correct, will change (and also change block_size to size_t)

aurel32 commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,104 +6,492 @@
+}
+
+/* flash write and verify */
+int _nvs_flash_write_ver(struct nvs_fs *fs, off_t offset, const void *data,
```
I know that `ver` means "verify", but "version" first came to my mind. I think it would make it more understandable to call it `_nvs_flash_write_cmp`, and would be consistent with `_nvs_flash_block_cmp`.

Laczen replied:
Yes, _cmp is more consistent, will change

aurel32 commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,104 +6,492 @@
+		}
+		offset += byte_cnt;
+		byte_cnt = 0;
+	}
```
All the cases in this function looks like error prone. For example if `write_block_size` is 2 bytes and the data length to be written below 128 bytes, `byte_cnt` will be 3. This mean the first call to `_nvs_flash_write_ver` will try to write 3 bytes, which is the wrong number for a 2-byte write block size.

Laczen replied:
This is an error, I will review this function once again to make it less error prone.

aurel32 commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,104 +6,492 @@
+	}
+	if (len_size == 2) {
+		_nvs_addr_advance(fs, &addr, sizeof(u16_t) + sizeof(u16_t));
+	}
```
It looks simpler to test for `len_size != 0xffff` and use the value of `len_size` directly instead of testing both cases.

aurel32 commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,104 +6,492 @@
+	/* if len_size = 0xffff just read from the address given as all should
+	 * be ff
+	 */
+	rc = _nvs_flash_read(fs, addr, data, len);
```
There you probably want to use `min(len, len_flash)`, to make sure the data from the next entry, or even from outside the NVS area is read. This could be a security issue when combined with missing input validation from the caller. I actually wonder if `_nvs_flash_read` should also limit the access to the NVS area only.

aurel32 commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,104 +6,492 @@
+					+ 2*sizeof(u16_t));
 }
 
+void _nvs_addr_advance(struct nvs_fs *fs, off_t *addr, size_t step)
```
I think all those private function that are not part of the API should just be marked as static in `nvs.c,` so that the compiler can do more aggressive optimizations. I just report that here, but that applies to many functions of this file.

aurel32 commented on include/nvs/nvs.h:
```diff
@@ -91,16 +91,17 @@ struct nvs_fs {
- * @retval 0 Success
- * @retval -ERRNO errno code if error
- */
-int nvs_clear(struct nvs_fs *fs);
```
I don't know zephyr enough to know about the API policy, but removing this function is breaking the API. I know there is also a file system format change, so that might be fine after all.

Laczen replied:
I will reintroduce nvs_clear, if the user would like to clear the flash area for some other purpose, nvs_clear should be there.

aurel32 reviewed (changes requested):
I have finally reviewed your changes, sorry for the delay. Overall the new flash format and the architecture of the code looks fine to me. I think it still needs some improvements in a few places, but nothing really big.

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,104 +6,492 @@
+size_t _nvs_put_length(u8_t *buf, size_t len)
 {
-	return entry->data_addr + _nvs_len_in_flash(fs, entry->len);
+	len &= 0x3fff;
```
Yes, I have introduced a define NVS_MAX_LEN (0x7ffd), 0xffff is returned if len> NVS_MAX_LEN.

nvlsianpu commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,104 +6,492 @@
+	len_size = _nvs_get_length(lenstr, &len_flash);
+
+	if (len_size == 1) {
+		_nvs_addr_advance(fs, &addr, sizeof(u8_t) + sizeof(u16_t));
```
why not `_nvs_addr_advance(fs, &addr, len_size + sizeof(u16_t));` instead of this 'if'?

Laczen replied:
Correct, changed

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,104 +6,492 @@
+	if (length_cnt == 1) {
+		ilenstr[0] = lenstr[0];
+	} else {
+		ilenstr[0] = lenstr[1];
```
Hi @nvlsianpu, thanks for all comments, the second length needs to be swapped because you will always read 2 bytes, but this can be data as well, so to make sure you get the correct length you are only sure about the last byte (this is definitely part of the length), so the last byte should be byte0 of the length.

nvlsianpu replied:
thank for elaboration.

nvlsianpu commented on subsys/fs/nvs/nvs.c:
```diff
@@ -162,214 +553,189 @@ void _nvs_entry_sector_advance(struct nvs_fs *fs)
+			_nvs_addr_advance(fs, &addr, _nvs_flash_size(fs,
+					  len_start));
+			rc = _nvs_flash_rd_len_id_end(fs, addr, &id, &len_end);
+			if (len_start != len_end) {
```
I think it would be worth to serve the case when this is about the last entry which was not entirely written to the flash (caused for instance by power down).

Laczen replied:
Hi @nvlsianpu, I am changing this routine a bit, fs->write_location is now updated everytime the length comparison is ok. So in the case of an entry that was not completely written the initialization routine will give a fs that is read-only but does not include the latest (invalid) entry. It is then possible to retrieve all information that is valid.

nvlsianpu replied:
What you are changing sound fine to me. Otherwise what I was asking might be added in a future PR as well. I think this is very likely to have the last entry corrupted for some devices. I would be great If nvs would can fix this in place by performing customized sector rotation.

nvlsianpu commented on subsys/fs/nvs/nvs.c:
```diff
@@ -389,400 +755,179 @@ int nvs_init(struct nvs_fs *fs, const char *dev_name, u32_t magic)
+	 * errors the sector is filled with smaller items.
+	 */
+	memset(buf, 0xff, NVS_BLOCK_SIZE);
+	while (len > (NVS_BLOCK_SIZE + _nvs_flash_size(fs, 0))) {
```
I understood that this fill procedure is introduced due the backward serach feature you had added?

Laczen replied:
Hi andrzej, also the first version of NVS has filling at the end of a sector. The length was even set to jump over the next sector header. However only the length was written to flash. Now ff's are written. The procedure has changed because of the variable length format, making the length parameter variable in size makes it impossible to reach certain sizes (e.g. on a system with byte write size a length of 127 bytes generates a fill block of 133 bytes; a length of 128 bytes generates a length of 136 bytes, so it is impossible to generate a block of size 134 and 135). This is of course more complicated for different write sizes, so I decided to avoid this difficulty by cutting it down to smaller blocks.

nvlsianpu commented on samples/subsys/nvs/src/main.c:
```diff
@@ -154,6 +155,30 @@ void main(void)
 	}
 
+	/* LONG_ID is used to store a larger dataset ,lets see if we can read
+	 * it from flash
```
note that "larger dataset" means those which are using the long format of length fields.

Laczen replied:
Yes

nvlsianpu commented:
@Laczen I put all comments I had.

Laczen commented:
Hi @aurel32 and @nvlsianpu, I have integrated your comments in my latest rework.

lemrey commented:
Good job guys! I'd suggest to make write verification optional, since some drivers already implement that.
In general, are there plans to make the flash program unit available at compile time? That'd simplify things a bit, I believe.

nvlsianpu commented on subsys/fs/nvs/nvs.c:
```diff
@@ -162,214 +503,215 @@ void _nvs_entry_sector_advance(struct nvs_fs *fs)
+		entry_len = get_wbs_aligned_entry_len(fs, len1);
+		_nvs_addr_advance(fs, &addr, entry_len);
+		rc = _nvs_flash_rd_len_id_end(fs, addr, &id, &len2);
+		if (len1 != len2) {
```
I think we still need an automatic mechanism for restore FS automatically after the power down while writing issue - it is something we really need for any device in the field. this comment is continuation of the https://github.com/zephyrproject-rtos/zephyr/pull/8141#discussion_r194684068 thread
cc @lemrey You opinion?

lemrey replied:
It is definitely desirable, the scenario is not uncommon.
I will comment on #8367.

Laczen commented:
Hi @lemrey, thanks for the comment. I think it's a good idea to make the verification optional. I will add this.

I don't quite get what you mean by making the flash program unit available at compile time. Could you explain this a bit more ?

lemrey commented:
Sure @Laczen ,
I wasn't referring to nvs in particular with that, but more generally about zephyr's flash subsystem.
I believe that flash libraries could benefit from knowing how large is the smallest unit of memory that can independently programmed (written). I think that is not possible at the moment, but it would spare some of the logic necessary for padding perhaps? :)

aurel32 commented:
> I believe that flash libraries could benefit from knowing how large is the smallest unit of memory that can independently programmed (written).

Even if you know it, it doesn't mean it is fixed. You can have a flash internal to the SoC with a write block size of 4 bytes, and an SPI flash with a write block size of 1.

lemrey commented:
You are right @aurel32 :/
I guess we'll have to live with that then!

lemrey commented on subsys/fs/nvs/nvs.c:
```diff
@@ -119,9 +457,7 @@ int _nvs_sector_is_used(struct nvs_fs *fs, off_t offset)
-		rc = flash_read(fs->flash_device,
-				fs->offset + offset + addr,
-				&buf, sizeof(buf));
+		rc = _nvs_flash_read(fs, offset + addr,	&buf, sizeof(buf));
```
Is there any reason why we a reading one byte at a time?
I would suggest to read a multiple, like a word, to reduce the number of calls.

lemrey commented on subsys/fs/nvs/nvs_priv.h:
```diff
@@ -18,55 +18,27 @@ extern "C" {
-			  * for the last entry in a sector
-			  */
-};
+#define NVS_BLOCK_SIZE 8
```
This is basically the maximum length that will be passed to the driver API in a single call.
I can definitely see a use for this, and I would be great to let the user see it menuconfig.

nvlsianpu commented on subsys/fs/nvs/nvs.c:
```diff
@@ -654,45 +660,104 @@ int nvs_reinit(struct nvs_fs *fs)
+		/* calculate last sector start */
+		_nvs_addr_decrease(fs, &addr, fs->sector_size);
+		/* erase the last sector */
+		rc = _nvs_flash_erase(fs, addr, fs->sector_size);
```
I don't understand why always physically the last sector is erased?

Laczen replied:
Hi Andrzej, the last added sector is erased. When all sectors are used this should be just before the entry sector.

nvlsianpu replied:
NVS uses N sectors, numbered form 0 to N-1. NVS stores entries in N-1 sectors, one sector should be always the scratch sectors.
Why once we got this "GC power down failure" you assume that it was the N-1nd is the properer scratch sector, not any else sector sector are used in round-robin fashion?

nvlsianpu replied:
@lemrey ^^ Did I miss something?

Laczen replied:
@nvlsianpu, NVS always keeps an empty sector, this sector is always just before the entry sector. When gc is started this empty sector is used to copy data from the entry sector. Now what happens if gc is incomplete: all sectors will be in use. We can find the entry sector by looking at the id's. As gc is always done on the entry sector we now what sector gc was doing, we also know where the data was going: to the latest added sector (which is just before the entry sector if all sectors are in use).
Maybe it's because my comment should be "erase last added sector" instead of "erase last sector".

lemrey replied:
My understanding is that we are always erasing the entry sector, but the entry sector moves forward so we are actually cyclying through all the sectors as one would expect. If there is no scratch area then the sector before the entry sector was the one used as scratch area, because an entry sector becomes the scratch area when the entry sector is moved forward.

nvlsianpu replied:
OK, it was my misunderstanding, because I forget and misinterpreted the  meaning of fs->entry_sector field.

Laczen commented:
@nvlsianpu, @aurel32, @lemrey, based upon the proposal of @lemrey to put the id at the end I am reworking nvs. There are two main advantages in putting the id at the end:
1. Read performance is increased by doing one less read from flash
2. By putting the id at the end it is possible to add a correction block after a incomplete write, the incomplete write is finished with a block that contains an id and length to jump over the invalid data. The id at the end allows detection of a "corrupt" block.

By adding the length in the beginning and the end also the sector_hdr and sector_end could be removed, this would create more storage space and simplify the code.

nvlsianpu commented on subsys/fs/nvs/nvs.c:
```diff
@@ -654,45 +660,104 @@ int nvs_reinit(struct nvs_fs *fs)
 
 	rc = 0;
+
 err:
```
label name is misleading in case it is successful exit.

nvlsianpu commented on subsys/fs/nvs/nvs.c:
```diff
@@ -654,45 +660,104 @@ int nvs_reinit(struct nvs_fs *fs)
+	fs->write_location = addr;
+	_nvs_addr_decrease(fs, &addr, fs->sector_size);
+	/* copy the OK data to temporary sector */
+	_nvs_flash_block_move(fs, addr, len1);
```
Is it up to the application that the touched data are still valid?

nvlsianpu replied:
Just have idea that we can return some value which signals that recovery was done and especial the last entry is not trusted. We can expected the NOR flash (so the SoC embedded flash) is reliable regards keeping its content - so it dose make sens to verify the content rare.

Laczen commented:
@nvlsianpu, @lemrey, I have updated NVS after your remarks. It should now be more robust against incomplete writes. I am still concerned about a disadvantage of NVS (which should also be the case for fcb): as flash wears errors can appear in flash, any error in the length will make it impossible to walk through the storage and thereby make all data unavailable.

dbkinder reviewed (approved):
LGTM

lemrey commented:
Hi @Laczen, thank you for contributing these improvements.
I had an idea regarding that problem; if my understanding is correct, an entry may not span across two sectors so the first bytes of a sector always contain an entry's header (if the sector is in use). It should be possible then, when an error is detected, to jump to the next sector directly and continue reading any remaining entries from there.

dbkinder reviewed (approved):
LTGM, thanks.

Laczen commented:
Hi @lemrey, you are right that the boundary between each sector is both an end of the previous sector as the start of a new sector. This could be used to get as much good items as possible from NVS. However it would be very difficult (travelling both forward and backwards). A different problem that appears is what do you do when you detect a write error, you cannot do a rewrite as you know the file system is broken.

I am exploring a different approach: suppose a separation is done between metadata (id and length) and data, the metadata is written from the sector end (in some kind of allocation table); the data is written from the sector start. If a data write fails a error entry is written in the allocation table. It is then possible just to restart the data write at the next available location, if it is OK the correct entry is written in the allocation table. It will take me some time to explore this further because I want minimal RAM and flash usage.

Laczen commented:
@lemrey, thanks for the comments. The idea I'm exploring now is:
1. Keep data and metadata in the same sector. Start metadata from the end, start data from the start. This requires keeping track of 2 addresses: one for the data and one for the metadata.
2. Keeping track of the space available is simple: it is just the difference between the (next) metadata address and the (next) data address.
3. The metadata would be: id, length, offset (in the sector) and a crc.
4. I would not start the metadata at the very end, but leave a space at the end to close a sector (this would direct to the last valid metadata of the sector). When a sector is closed this could be detected because it is different from 0xff.
5. The metadata would include a link to the data, so in gc the data could be copied directly and then a new metadata would be written.
6. I would always first write the data and then the metadata, this would allow the metadata to invalidate a unsuccessful write and makes it easy to correct incomplete writes.

lemrey commented:
@Laczen , yeah keeping track of the space shouldn't be a problem, but it would require RAM per-sector instead of a constant; that's acceptable in my opinion.

I don't see why to keep the offset of the entry data in the metadata though. To lookup an entry we'd still have to "jump" all the way through, we might as well just accumulate that offset from the previous entries' lengths. This would let us copy the metadata as-is during GC and save some space in flash. Also, if we keep an offset per-sector why should we "close" a sector if an entry doesn't fit? We could just leave it as-is and use it to store any entry that might fit in the future.

Laczen commented:
@lemrey, once again thanks for the comment.

You are right, keeping the offset and length is redundant, but this would eliminate the difficulty when an entry in the metadata would be invalid. Then it no longer possible to jump through correctly. If the metadata does contain a link to the data location each of the metadata entries could have a problem but it would not influence other metadata/data entries.

Regarding the closing of the sectors, I have been thinking about this also. Closing the sectors has the following advantages:
1. RAM usage: only two variables are required: the write location of data and metadata in the active sector.
2. Easier initialization because it is only needed to find the current in use sector.
3. Keeping track of history items is possible because earlier sectors contain earlier information (otherwise it is very complex because a rewrite of an item with smaller size might get in a previous sector).
As a solution for the wasted space at the end of a sector i would later extend it with the possibilities to save the data in multiple parts.

lemrey commented:
You made some very good points @Laczen, I think you are on the right track :)

nvlsianpu commented:
@Laczen, @carlescufi  So far changes look good me. I will be on vacation next to weeks - so I can say that I'm surre that @lemrey review will be enough to accept and merge this PR.

Laczen commented:
Hi @carlescufi, this is an improved version of NVS to provide better support on systems with write block size > 1, and improved robustness against incomplete writes.
I am working on a even more robust version that would be more robust against flash read errors. It will still take some time before this is ready. If you don't mind waiting a bit more this PR can be closed and I will create a new PR when the "final" version is ready.

nvlsianpu commented:
@Laczen Better to keep this within this PR as we have discussion story here.

Laczen committed:
subsys: fs/nvs: Rewrite for improved robustness

On flash NVS was stored one entry after another including the metadata
of each entry. This has the disadvantage that when an incomplete write
is performed (e.g. due to power failure) the complete sector had to be
rewritten to get a completely functional system.

The present rewrite changed the storage in flash of the data. For each
sector the data is now written as follows: the data itself at the
beginning of the sector (one after the other), the metadata (id, length,
data offset in the sector, and a crc of the metadata) is written from
the end of the sector. The metadata is of fixed size (8 byte) and for
a sector that is completely occupied a metadata entry of all zeros is
used.

Writing data to flash always is done by:
1. Writing the data,
2. Writing the metadata.

If an incomplete write is done NVS will ignore this incomplete write.

At the same time the following improvements were done:
1. NVS now support 65536 sectors of each 65536 byte.
2. The sector size no longer requires to be a power of 2 (but it
still needs to be a multiple of the flash erase page size).
3. NVS now also keeps track of the free space available.

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

Laczen commented:
@nvlsianpu, @lemrey, @aurel32, to improve robustness against incomplete writes I have rewritten NVS with a completely different layout on flash. It now stores the data at the beginning of a sector and the metadata (id, length and a offset to the data) at the end of a sector. This way NVS should always be readable in case of a incomplete write (either data or metadata).
@dbkinder, I still have to update the documentation to reflect the changes.

nvlsianpu commented on subsys/fs/nvs/nvs.c:
```diff
@@ -10,161 +10,126 @@
+	offset += addr & 0xffff;
 
-/* end basic calculation routines */
+	rc = flash_read(fs->flash_device, offset, data, len);
```
Is it ensured that read operations are aligned to block-size? It is not expected form a flash driver that it support byte alignment while reading.

nvlsianpu commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,814 +6,764 @@
-		rc = nvs_flash_read(fs, entry->data_addr + i, buf, sizeof(buf));
+	/* Normally addr points to a location that does not contain all ff, so
+	 * ate_size needs to be substracted to get ate_wra. However there is an
+	 * exveption, when addr is the very end of a sector; then we keep
```
exveption -> exception

nvlsianpu commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,814 +6,764 @@
+		if (!empty_len) {
+			break;
+		}
+		_nvs_flash_cmp_const(fs, fs->data_wra, 0xff, empty_len);
```
rc = _nvs_flash_cmp_const(...

nvlsianpu commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,814 +6,764 @@
+	}
+	if (rc) {
+		/* fs->ate_wra not at sector end */
+		fs->ate_wra -= ate_size;
```
shouldn't nvs iterates more for seek ate_wra? this assume that ate_wra referencing the second ate in a sector.

Laczen replied:
no ate_wra is found by _nvs_prev_ate(), more iterations are not needed. ate_wra is not referencing the second ate in a sector.

nvlsianpu replied:
ok

nvlsianpu commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,814 +6,764 @@
-		goto err;
+
+	if (fs->free_space < (data_size + ate_size)) {
+		return -ENOSPC;
```
revert free space value before exit as write didn't happen - a smaller entry might fit.

Laczen replied:
correct, wil change this

nvlsianpu commented on subsys/fs/nvs/nvs.c:
```diff
@@ -822,105 +772,62 @@ int nvs_delete(struct nvs_fs *fs, u16_t id)
 	}
-	rc = entry.len;
+
+	return wlk_ate.len;
```
In API documentation pleas rephrase `Any other value, indicates an error.` -> `Any negative value, indicates an error.`
This is applicable to the nvs_read() as well.

nvlsianpu replied:
This is fixed.

nvlsianpu commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,814 +6,764 @@
+		}
+		/* stop after stepping through all sectors */
+		if (*addr == fs->ate_wra) {
+			break;
```
I don't understand ho we come to this case as fs->ate_wra refer to place before newest ate metadata.

Laczen replied:
This happens when we have completely gone through nvs in reverse direction.

nvlsianpu reviewed (changes requested):
I reviewed re-implementation complacently.

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,814 +6,764 @@
+	u8_t buf[fs->write_block_size];
 
+	offset = fs->offset;
+	offset += fs->sector_size * (addr >> 16);
```
Hi @nvlsianpu, I am not sure what you mean by this comment. The address definition that is used in nvs is: high 2 bytes are the sector number, low 2 bytes are the offset in the sector. The calculation is mapping the address to an offset in flash. All address masks used are either to find the sector (addr >> 16) or the offset (addr & 0xffff).

nvlsianpu replied:
That I understood well. What I mens to use predefined mask for extracting these numbers/addresses, something like that:
```
#define  _SECTOR_NUM_MASK 0xFFFF0000
#define  _SECTOR_NUM_SHIFT 16
#define  _OFF_MASK 0x0000FFFF
```

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,814 +6,764 @@
-	k_mutex_unlock(&fs->nvs_lock);
-	return rc;
-}
+	fs->ate_wra = addr;
```
addr has been changed by _nvs_prev_ate()

nvlsianpu replied:
ok

Laczen commented on subsys/fs/nvs/nvs.c:
```diff
@@ -6,814 +6,764 @@
-	if ((id == NVS_ID_EMPTY) ||
-	    (id == NVS_ID_SECTOR_END) ||
-	    (len > fs->max_len) ||
+	if ((id == NVS_ID_EMPTY) || (len > (fs->sector_size - 2*ate_size)) ||
```
I think leaving spaces around * makes it confusing, I am using spaces for +/- but not for *.

nvlsianpu replied:
I see you point.

Laczen commented:
Hi @nvlsianpu, I think all your remarks have been addressed in the latest commit.

nvlsianpu commented:
@Laczen I'm happy witch changes. Can you fix shippable, merge conflict and clean git history (squash some commit, especial these which align code to review request.

Appart form that the code is in expected shape to be approved.

Laczen commented:
@nvlsianpu, fixed shippable and merge conflict (was result of removing no longer valid configuration option NVS_PROTECT_FLASH, I will remove this option in a later PR).

nvlsianpu commented:
@nashif Ready to be merged IMO.

nvlsianpu commented:
@nashif, @aurel32 - have you still any pending comments on the code - It is ready since few weeks.
@lemrey - has not as far I know.
@galak, @carlescufi  - can we merge this or we should wait for @nashif

aurel32 commented:
Hi @Laczen. Sorry to not come back earlier about this, I have been busy with real life.

I gave a look to your new implementation, and I like the new design of not mixing the data and the metadata. Thanks also for keeping a CRC. Overall it looks good to me. I have however found a few issues and small things that can be improved. The main one is that it doesn't work when block size != 1. I will send a pull request to fix it. I will also prepare a pull request for the small improvements.

I have still a few things that I don't fully understand, I hope you don't mind I ask them there:

- There are 3 places with this code:
`block_size = NVS_BLOCK_SIZE & ~(fs->write_block_size - 1);`
I am not fully sure what it is supposed to do. Moreover it will return 0 if write_block_size is bigger than NVS_BLOCK_SIZE and will end-up in an endless loop.

- I am not fully sure about the free space accounting. In my case with 2x2048-byte sectors, the free space is always 7c8 aka 1992 bytes at start, independently of the data already present. In addition I think there is an issue in the following part of _nvs_flash_wrt:
```
        rc = _nvs_flash_data_wrt(fs, data, len);
        if (rc) {
                return rc;
        }
        rc = _nvs_flash_ate_wrt(fs, &entry);
        if (rc) {
                return rc;
        }

        if (len != 0) {
                fs->free_space -= _nvs_al_size(fs, len);
                fs->free_space -= ate_size;
        }
```
When len is 0, the ate is still written, so the free space is reduced. I am also not sure it's a good idea to call `_nvs_flash_data_wrt`. Even it if doesn't write anything, it still unsets the flash protection and sets it back.

- `nvs_clear` calls `_nvs_flash_erase_sector` on all sectors which errors out if the sector is already erased. It means that all sectors might not been erased.

Laczen commented:
Hi @aurel32, thanks for the fix when block size !=1, I don't understand how
I missed this. I have also been busy doing other things.

Regarding the calculation of block_size, this is only to protect for silly
definitions of NVS_BLOCK_SIZE, e.g. setting it to 12 on a system with
write_block_size 8. You are correct that there is a problem when
NVS_BLOCK_SIZE is smaller than write_block_size, but I don't know how to
fix that. Since NVS can be used on external flash, internal flash, ...
there is no way of figuring out in advance what a good setting of
NVS_BLOCK_SIZE is. I set it to 8 because I expect that for the moment that
are no systems with write_block_size bigger than 8.

Regrading the free space calculation: the free space calculation should
return how much data can be written to the system (possibly after garbage
collection), it is not an indication of how much flash is used at the
moment. Deleted items will not be copied when doing gc, so their data and
also ate should not be counted. On your system I would expect it to report
0x7d8, as NVS will always need two extra ate's for pointing to the last
entry and a sector closing ate. I need to go through the code again to see
why there is space reserved for two extra ate's (possibly to always allow a
delete).

Regarding the call of _nvs_flash_data_wrt(), the catch of len=0 would be
good to handle in the aligned write, than the flash protection unset and
set can be avoided.

Regarding nvs_clear I need to check the code again.

Op do 23 aug. 2018 om 23:37 schreef aurel32 <notifications@github.com>:

> Hi @Laczen <https://github.com/Laczen>. Sorry to not come back earlier
> about this, I have been busy with real life.
>
> I gave a look to your new implementation, and I like the new design of not
> mixing the data and the metadata. Thanks also for keeping a CRC. Overall it
> looks good to me. I have however found a few issues and small things that
> can be improved. The main one is that it doesn't work when block size != 1.
> I will send a pull request to fix it. I will also prepare a pull request
> for the small improvements.
>
> I have still a few things that I don't fully understand, I hope you don't
> mind I ask them there:
>
>    -
>
>    There are 3 places with this code:
>    block_size = NVS_BLOCK_SIZE & ~(fs->write_block_size - 1);
>    I am not fully sure what it is supposed to do. Moreover it will return
>    0 if write_block_size is bigger than NVS_BLOCK_SIZE and will end-up in an
>    endless loop.
>    -
>
>    I am not fully sure about the free space accounting. In my case with
>    2x2048-byte sectors, the free space is always 7c8 aka 1992 bytes at start,
>    independently of the data already present. In addition I think there is an
>    issue in the following part of _nvs_flash_wrt:
>
>         rc = _nvs_flash_data_wrt(fs, data, len);
>         if (rc) {
>                 return rc;
>         }
>         rc = _nvs_flash_ate_wrt(fs, &entry);
>         if (rc) {
>                 return rc;
>         }
>
>         if (len != 0) {
>                 fs->free_space -= _nvs_al_size(fs, len);
>                 fs->free_space -= ate_size;
>         }
>
> When len is 0, the ate is still written, so the free space is reduced. I
> am also not sure it's a good idea to call _nvs_flash_data_wrt. Even it if
> doesn't write anything, it still unsets the flash protection and sets it
> back.
>
>    - nvs_clear calls _nvs_flash_erase_sector on all sectors which errors
>    out if the sector is already erased. It means that all sectors might not
>    been erased.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zephyrproject-rtos/zephyr/pull/8141#issuecomment-415579175>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAjVvNmiB0x5tY17ZInmPDvH_p-McaMSks5uTyCbgaJpZM4UYs0O>
> .
>

aurel32 commented:
> Hi @aurel32, thanks for the fix when block size !=1, I don't understand how I missed this. I have also been busy doing other things. 

No problem, this kind of things happen.

> Regarding the calculation of block_size, this is only to protect for silly definitions of NVS_BLOCK_SIZE, e.g. setting it to 12 on a system with write_block_size 8. You are correct that there is a problem when NVS_BLOCK_SIZE is smaller than write_block_size, but I don't know how to fix that.

Ok, now that I understand the use case, I'll try to think about a solution. As you said very few systems should have a write block size bigger than 8 bytes.

> Regrading the free space calculation: the free space calculation should return how much data can be written to the system (possibly after garbage collection), it is not an indication of how much flash is used at the moment.

Thanks for the explanation, I'll look at that part again with that in mind.

> Regarding the call of _nvs_flash_data_wrt(), the catch of len=0 would be good to handle in the aligned write, than the flash protection unset and set can be avoided.

Ok, I'll try to do that and include that in PR #9601.

Laczen commented:
Hi @aurel32, I have had some time to look into the problem of the free space calculation as well as nvs_clear.

I have not been able to reproduce your numbers for the free space calculation. When nvs is initialized (with a sector of 2048 byte) it reports a free space of 2040 bytes, and the alloc wra is 0x7f8.

For nvs_clear() I also do not see it error out on already erased sectors, the compare with 0xff returns a zero and _nvs_flash_erase_sector() returns a 0. It just does not do a erase (because it is not needed).

aurel32 commented:
Hi @Laczen, thanks for checking my points on `nvs_clear` and free space calculation. I have tried to look at both of them in details, and then both looks fine. I probably test them wrongly last time, sorry about that.
